### PR TITLE
Update composer packages; Bump PHP to 7.4 and WP to 5.3

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary

- Update composer dev packages to the latest version.
- Bump PHP to 7.4
- Bump minimum WP version to 5.3 
- Update workflows to use globally installed PHPUnit.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -20,23 +20,22 @@
     "sabberworm/php-css-parser": "dev-master#cc791ad"
   },
   "require-dev": {
-    "automattic/vipwpcs": "^2.2",
+    "automattic/vipwpcs": "^2.0",
```
Comment (westonruter) [inline composer.json]: 3.0.0 was just released: https://github.com/Automattic/VIP-Coding-Standards/releases/tag/3.0.0

Comment (thelovekesh) [reply composer.json]: Yes just sending a commit.

File: tests/php/test-class-amp-comments-sanitizer.php [hunk 2]
```diff
@@ -218,7 +218,7 @@ public function test_ampify_threaded_comments( $ampify_comment_threading, $comme
 			$this->assertInstanceOf( Element::class, $span );
 			$this->assertTrue( $span->hasAttribute( Amp::BIND_DATA_ATTR_PREFIX . 'text' ) );
 
-			$comment_reply_links = $dom->xpath->query( '//a[ @data-commentid and @data-postid and @data-replyto and @data-respondelement and contains( @class, "comment-reply-link" ) ]' );
+			$comment_reply_links = $dom->xpath->query( '//a[ @data-commentid and @data-postid and @data-respondelement and contains( @class, "comment-reply-link" ) ]' );
```
Comment (thelovekesh) [inline tests/php/test-class-amp-comments-sanitizer.php:217]: `data-replyto` was added in WP 5.5 - https://github.com/WordPress/wordpress-develop/commit/251fd685dd932200087ad04fa49c598ef9657cda

File: includes/sanitizers/class-amp-comments-sanitizer.php [hunk 3]
```diff
@@ -218,7 +218,7 @@ protected function ampify_threaded_comments( Element $comment_form ) {
 		}
 
 		// Update comment reply links to set the reply state.
-		$comment_reply_links = $this->dom->xpath->query( '//a[ @data-commentid and @data-postid and @data-replyto and @data-respondelement and contains( @class, "comment-reply-link" ) ]' );
+		$comment_reply_links = $this->dom->xpath->query( '//a[ @data-commentid and @data-postid and @data-respondelement and contains( @class, "comment-reply-link" ) ]' );
```
Comment (thelovekesh) [inline includes/sanitizers/class-amp-comments-sanitizer.php:221]: See above comment - https://github.com/ampproject/amp-wp/pull/7613#discussion_r1317685805

File: .github/workflows/build-test-measure.yml [hunk 4]
```diff
@@ -101,8 +101,6 @@ jobs:
 
       - name: Install Node dependencies
         run: npm ci
-        env:
-          CI: true
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml:105]: Why removing this?

File: .github/workflows/build-test-measure.yml [hunk 5]
```diff
@@ -508,55 +498,20 @@ jobs:
...
-            composer require --dev --ignore-platform-reqs phpunit/phpunit:"9.3.*" --with-dependencies
-          elif [[ $WP_VERSION == "latest" || $WP_VERSION == "trunk" || $PHP_VERSION == "8.1" || $PHP_VERSION == "8.2" ]]; then
-            echo "Installing latest version of PHPUnit"
-            composer update --ignore-platform-reqs --no-interaction --no-scripts yoast/phpunit-polyfills --with-dependencies
-          fi
-        env:
-          WP_VERSION: ${{ matrix.wp }}
-          PHP_VERSION: ${{ matrix.php }}
+          rm -rf vendor/phpunit
+          composer dump-autoload -o
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml:519]: This is better than doing `composer remove phpunit/phpunit`?

File: .github/workflows/build-test-measure.yml [hunk 6]
```diff
@@ -646,26 +609,26 @@ jobs:
       fail-fast: false
       matrix:
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml:619]: Should 8.3 be added experimentally? I understand that WordPress `trunk` is now compatible.

File: .github/workflows/build-test-measure.yml [hunk 7]
```diff
@@ -646,26 +609,26 @@ jobs:
...
         coverage: [false]
-        php: ['7.4', '7.3', '7.2', '7.1']
+        php: ['7.4']
         wp: ['latest']
         include:
+          - php: '8.2'
+            wp: 'trunk'
+            experimental: true
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml:631]: Does this need to be `experimental`?

```suggestion
```

File: .github/workflows/build-test-measure.yml [hunk 8]
```diff
@@ -646,26 +609,26 @@ jobs:
...
             wp: 'latest'
-            coverage: false
 
-          # TODO: Enable this once codecov upload merging works as expected.
-          # - php: '7.4'
-          #  wp: 'latest'
-          #  coverage: true
+          # @TODO: Enable code coverage for feature tests if codecov supports it.
+          - php: '7.4'
+            wp: 'latest'
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml]: Is this not a duplicate?

File: .github/workflows/build-test-measure.yml [hunk 9]
```diff
@@ -711,21 +674,20 @@ jobs:
...
+      # @TODO: Revisit this if we want to enable code coverage for feature tests.
+      # - name: Update PHPUnit to get latest php-code-coverage library
+      #   if: ${{ matrix.coverage == true }}
+      #   # phpdocumentor/reflection has to be removed as it makes use of an outdated dependency.
+      #   # phpunit/phpunit has to be updated as the one in use provides an older version of phpunit/php-code-coverage,
+      #   # but we need the v9.x branch.
+      #   # It cannot be removed, as it is a requirement of wp-cli/wp-cli-tests as well.
+      #   run: |
+      #     composer remove --dev phpdocumentor/reflection
+      #     composer require --dev --ignore-platform-reqs --update-with-all-dependencies phpunit/phpunit
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml:699]: Code coverage wasn't working for feature tests anyway, right?

File: tests/php/test-class-amp-comments-sanitizer.php (hunk 2 above)
Comment (westonruter) [reply tests/php/test-class-amp-comments-sanitizer.php:217]: I'm confused. Why does it need to be removed?

File: tests/php/test-class-amp-meta-box.php [hunk 10]
```diff
@@ -121,6 +121,11 @@ public function test_enqueue_admin_assets() {
 	 * @covers ::enqueue_block_assets()
 	 */
 	public function test_enqueue_block_assets() {
+		// Block validation script uses features only available beginning with WP 5.6.
+		if ( version_compare( get_bloginfo( 'version' ), '5.6', '<' ) ) {
+			$this->markTestSkipped( 'The block validation script is only available in WordPress 5.6 and above.' );
+		}
```
Comment (westonruter) [inline tests/php/test-class-amp-meta-box.php:127]: Why is this needed now if previously we were testing back to 4.9?

File: .github/workflows/build-test-measure.yml (hunk 4 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:105]: This is now implemented at the platform level. GitHub runners already declare this env var before starting a job.

File: .github/workflows/build-test-measure.yml (hunk 5 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:519]: `composer remove phpunit/phpunit` is unable to run it as PHPUnit is required by many other packages, and I am unable to find a way to ignore PHPUnit in the class map so it doesn't collide with global PHPUnit.

File: .github/workflows/build-test-measure.yml (hunk 9 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:699]: Yes as we are unable to merge coverage reports. I need to check if our coverage logic is working as expected or not.

File: tests/php/test-class-amp-meta-box.php (hunk 10 above)
Comment (thelovekesh) [reply tests/php/test-class-amp-meta-box.php:127]: Because we are checking for dependency support in `enqueue_block_assets()`

https://github.com/ampproject/amp-wp/blob/956c8d298764e4b22b936347d7c8b591b2fadfc2/includes/admin/class-amp-post-meta-box.php#L220-L224

File: tests/php/test-class-amp-comments-sanitizer.php (hunk 2 above)
Comment (thelovekesh) [reply tests/php/test-class-amp-comments-sanitizer.php:217]: Because xpath query with `@data-replyto` will not work for WP < 5.5 as `data-reply` attribute was added in WP 5.5. Hence it will not get comment reply links on WP 5.3 and 5.4

File: tests/php/src/Validation/ScannableURLProviderTest.php [hunk 11]
```diff
@@ -425,9 +426,21 @@ public function test_get_taxonomy_links() {
 				);
 			}
 
+			$post = self::factory()->post->create();
+
+			// `wp_pattern_category` is a special taxonomy that is registered for categorizing patterns.
```
Comment (thelovekesh) [inline tests/php/src/Validation/ScannableURLProviderTest.php:429]: Added in recent GB release - https://github.com/WordPress/gutenberg/pull/53163

File: .github/workflows/build-test-measure.yml (hunk 7 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:631]: PHP 8.2 is still in beta for WP 6.3. Should we remove it?

Comment (westonruter) [reply .github/workflows/build-test-measure.yml:631]: If the tests currently pass in 6.2 I'd say remove it.

File: .github/workflows/build-test-measure.yml (hunk 8 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml]: Removed.

File: .github/workflows/build-test-measure.yml (hunk 7 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:631]: Tests are passing with PHP 8.2 on WP 6.2, so experimental flag.

File: composer.json [hunk 12]
```diff
@@ -20,23 +20,22 @@
     "sabberworm/php-css-parser": "dev-master#cc791ad"
   },
   "require-dev": {
-    "automattic/vipwpcs": "^2.2",
+    "automattic/vipwpcs": "^3.0",
     "civicrm/composer-downloads-plugin": "^3.0",
-    "dealerdirect/phpcodesniffer-composer-installer": "0.7.2",
```
Comment (westonruter) [inline composer.json:25]: No longer necessary?

File: composer.json [hunk 13]
```diff
@@ -89,10 +88,6 @@
       }
     },
     "patches": {
-      "phpunit/phpunit-mock-objects": {
-        "Fix ReflectionParameter warnings on PHP 8": "patches/phpunit-mock-objects.patch",
-        "Remove 'match' keyword from uses <https://github.com/sebastianbergmann/phpunit/issues/4373>": "patches/remove-match-keyword.patch"
```
Comment (westonruter) [inline composer.json:94]: These files can be deleted from the repo now, correct?

File: composer.json (hunk 12 above)
Comment (thelovekesh) [reply composer.json:25]: WPCS handles it well as they need to keep few standards in installation path.

File: composer.json (hunk 13 above)
Comment (thelovekesh) [reply composer.json:94]: Yes and I have removed them as well.

File: tests/php/src/Validation/ScannableURLProviderTest.php [hunk 14]
```diff
@@ -425,9 +426,21 @@ public function test_get_taxonomy_links() {
...
+			// It is only associated with the `wp_block` post type, so we need to create a `wp_block` post.
+			if ( 'wp_pattern_category' === $taxonomy ) {
+				$post = self::factory()->post->create(
```
Comment (westonruter) [inline tests/php/src/Validation/ScannableURLProviderTest.php]: This post is overriding the `$post` created just before. Should the above `$post` creation be put in an `else`?

File: tests/php/validation/test-class-amp-validation-manager.php [hunk 15]
```diff
@@ -1215,8 +1212,10 @@ function ( $sources ) {
 
 			'wp_editor_has_scripts_attributed'           => [
 				function () {
-					if ( version_compare( get_bloginfo( 'version' ), '5.2', '<=' ) ) {
-						$this->markTestSkipped( 'The script ID attribute was only added to inline scripts in WP 5.2.' );
+					if ( version_compare( get_bloginfo( 'version' ), '5.5', '<' ) ) {
+						// @see <https://github.com/WordPress/wordpress-develop/commit/d241ab3b55b06fed59e7ea03318e9f872399b390>
+						//@see <https://github.com/WordPress/wordpress-develop/commit/d241ab3b55b06fed59e7ea03318e9f872399b390>
```
Comment (westonruter) [inline tests/php/validation/test-class-amp-validation-manager.php]:
```suggestion
						// @see <https://github.com/WordPress/wordpress-develop/commit/d241ab3b55b06fed59e7ea03318e9f872399b390>
```

Comment (westonruter) [reply tests/php/validation/test-class-amp-validation-manager.php]: Actually, this line is duplicated, so:

```suggestion
```

File: tests/php/validation/test-class-amp-validation-manager.php [hunk 16]
```diff
@@ -1215,8 +1212,10 @@ function ( $sources ) {
 
 			'wp_editor_has_scripts_attributed'           => [
 				function () {
-					if ( version_compare( get_bloginfo( 'version' ), '5.2', '<=' ) ) {
```
Comment (westonruter) [inline tests/php/validation/test-class-amp-validation-manager.php:1218]: Strange that 5.2 was the version here when 5.5 is the version it was introduced in.

File: tests/php/src/Validation/ScannableURLProviderTest.php (hunk 14 above)
Comment (thelovekesh) [reply tests/php/src/Validation/ScannableURLProviderTest.php]: Done in https://github.com/ampproject/amp-wp/pull/7613/commits/9b700fffab2bc73a3c19cd9dce10777c23337d61

File: src/Validation/ScannableURLProvider.php [hunk 17]
```diff
@@ -307,7 +310,7 @@ private function get_posts_by_type( $post_type, $offset = null ) {
 		}
 		$query = new WP_Query( $args );
 
-		return $this->get_posts_that_support_amp( $query->posts );
+		return $this->get_posts_that_support_amp( $query->posts, $posts_to_exclude );
```
Comment (westonruter) [inline src/Validation/ScannableURLProvider.php]: Instead of updating the `get_posts_that_support_amp` method with this new parameter, it would seem better to me to just do the filtering here in this method:

```suggestion
		$posts = array_values( 
			array_filter( 
				$query->posts, 
				static function ( $post ) use ( $posts_to_exclude ) { 
					return ! in_array( $post->ID, $posts_to_exclude, true ); 
				} 
			) 
		);
		return $this->get_posts_that_support_amp( $posts );
```

Comment (westonruter) [reply src/Validation/ScannableURLProvider.php]: Hold on, this can be simplified further since we're getting IDs:

```suggestion
		return $this->get_posts_that_support_amp( array_diff( $query->posts, $posts_to_exclude ) );
```

**Comment by westonruter** [comment]
I just had a concerning thought that this would cause problems for Web Stories. But I see that it already [requires PHP 7.4](https://github.com/GoogleForCreators/web-stories-wp/blob/1ca31eb1d739e67463abd8921cc00ddc1fc48a4c/web-stories.php#L14). So we're all good there.

File: src/Validation/ScannableURLProvider.php (hunk 17 above)
Comment (thelovekesh) [reply src/Validation/ScannableURLProvider.php]: I thought since we already filter these post IDs and `array_diff` is considered slow, completing the task in a single run would be performant. However, any performance impact should be negligible here as the array size will be 100 only.

Comment (westonruter) [reply src/Validation/ScannableURLProvider.php]: So we can go with `array_diff()` instead? It's less code and doens't change the resulting API.

**Review by westonruter — APPROVED** [review]
Awesome!
